apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js

```diff
@@ -156,31 +161,40 @@ router.get(
       const groupConfig = await groupAssessmentHelper.getGroupConfig(res.locals.assessment.id);
       res.locals.groupConfig = groupConfig;
 
-      // Check whether the user is currently in a group in the current assessment by trying to get a group_id
-      const groupId = await groupAssessmentHelper.getGroupId(
-        res.locals.assessment.id,
-        res.locals.user.user_id,
+      res.locals.notInGroup = false;
+      const groupInfo = await groupAssessmentHelper.getGroupInfo(
+        res.locals.assessment_instance.group_id,
+        groupConfig,
       );
+      res.locals.groupSize = groupInfo.groupSize;
+      res.locals.groupMembers = groupInfo.groupMembers;
+      res.locals.joinCode = groupInfo.joinCode;
+      res.locals.groupName = groupInfo.groupName;
+      res.locals.start = groupInfo.start;
+      res.locals.rolesInfo = groupInfo.rolesInfo;
+      res.locals.used_join_code = req.body.used_join_code;
 
-      if (groupId === null) {
-        throw error.make(403, 'Not a group member', res.locals);
-      } else {
-        res.locals.notInGroup = false;
-        const groupInfo = await groupAssessmentHelper.getGroupInfo(groupId, groupConfig);
-        res.locals.groupSize = groupInfo.groupSize;
-        res.locals.groupMembers = groupInfo.groupMembers;
-        res.locals.joinCode = groupInfo.joinCode;
-        res.locals.groupName = groupInfo.groupName;
-        res.locals.start = groupInfo.start;
-        res.locals.rolesInfo = groupInfo.rolesInfo;
-        res.locals.used_join_code = req.body.used_join_code;
+      if (groupConfig.has_roles) {
+        const result = await groupAssessmentHelper.getAssessmentPermissions(
+          res.locals.assessment.id,
+          res.locals.user.user_id,
+        );
+        res.locals.userCanAssignRoles = result.can_assign_roles_at_start;
 
-        if (groupConfig.has_roles) {
-          const result = await groupAssessmentHelper.getAssessmentPermissions(
-            res.locals.assessment.id,
-            res.locals.user.user_id,
-          );
-          res.locals.canViewRoleTable = result.can_assign_roles_at_start;
+        res.locals.user_group_roles =
+          groupInfo.rolesInfo?.roleAssignments?.[res.locals.authz_data.user.uid]
+            ?.map((role) => role.role_name)
+            ?.join(', ') || 'None';
+        // Get the role permissions. If the authorized user is a staff member, then they have all permissions.
```

Comment:
Maybe this is a dumb question: where is "they have all permissions" implemented? Are we just checking `res.locals.authz_data....` in the EJS template? Surely we don't want *all* staff members to have all permissions (that is, a student data viewer shouldn't be able to take actions)?

Reply:
The comment was misleading, but the implementation is correct. If a user has course instance view permission, then role restrictions don't apply. This is rephrased in 4ccba7e.